qiskit/aqua/operators/legacy/weighted_pauli_operator.py

```diff
@@ -103,7 +103,8 @@ def to_opflow(self, reverse_endianness=False):
             pauli = Pauli.from_label(str(p)[::-1]) if reverse_endianness else p
             # Adding the imaginary is necessary to handle the imaginary coefficients in UCCSD.
             # TODO fix these or add support for them in Terra.
-            pauli_ops += [PrimitiveOp(pauli, coeff=np.real(w) + np.imag(w))]
+            coeff = np.real(w) if not np.iscomplex(w) else w
```

Comment:
So we take the real part of w if it is not complex and if its complex pass it through. I know the issue was around not supporting complex numbers but the logic for taking the real part if its not complex seems odd to me.

Reply:
Yes, workaround until https://github.com/Qiskit/qiskit-terra/pull/5192 is merged.

Reply:
It still seems backwards to me. If coeff should never be complex for now shouldn't it be
`coeff = np.real(w) if np.iscomplex(w) else w` without the `not`?

Reply:
`w` can be a complex number, but ParameterExpression does not allow complex numbers. 
So some tests are not passed until `coeff` is a real number. 
The value should not be changed here. This is the type problem.

Reply:
I am still missing something - it is truly the behavior of the first line below we want? (as per the code with `not`)
```
>>> w =9+0.8j; np.real(w) if not np.iscomplex(w) else w
(9+0.8j)
>>> w =9+0.8j; np.real(w) if np.iscomplex(w) else w
9.0
>>> 
```

Reply:
I think the above is correct. I don't want to change the value for #1311.

Reply:
Ok, it just seems quite odd to me taking the real part of w when it is known to not be complex (i.e when it is already real)

Reply:
I see you changed the code from `if not np.iscomplex(w)` to `if np.isreal(w)` which logically is the same.

What still is puzzling to me is that, in either case, the logic still seems that it could be simply `coeff = w` 
In the expression there is takes the real(w) part of w if np.isreal(w), which seems to be no-op and otherwise sets coeff to w which could be a complex still.

I feel I must be missing something something here!

As I said if the code was
```
if w is complex
  coeff = real part of w
else
  coeff  = w
```
i.e
```
coeff = np.real(w) if np.iscomplex(w) else w
```
I would understand it - with the logic of the if reversed i.e. `not np.iscomplex(w)` or `np.isreal(w)` I still don't get it as I thought the intent was that since Terra did not support complex coeff in the parameters for now we could not pass complex on further - that was what the code did before by somehow making a real number out of the sum of the real and imaginary parts.

Reply:
I don't want to change an value but only a TYPE. The previous code breaks the value. e.g. if w is `1 + 1j `, np.real(w) + np.imag(w) is 1 + 1 = 2. This causes the bug.
If w is `1 + 1j`, the type of w must be a `complex`. But if w is `1.2`, the type of w can be converted into `float`.

Reply:
You mean if you have `1.2 + 0j` as  complex the intent is to simply make it a float since it has a 0 imaginary part. I guess its a simplification that can be done but if the coeff accepts complex anyway is that really needed?

My understanding of why that original logic was there is that in the UCCSD variational form it creates evolutionary circuits that ended up with complex coeffs. These coeffs end up in the parameters of the circuit of the variational form where Terra does not accept them since they could be complex depending on specific UCSSD form. Hence they were altered in some way - I seem to recall in the past we just took the real parts. Here when converting to opflow it seems the author summed the real and imag parts.

Reply:
This type conversion is necessary if someone wants to do #1311. And this spec is confusing for users. This is workaround until Terra's ParameterExpression allows complex number. So if the PR of Terra is merged, I will delete this type conversion. 

BTW if this original logic is necessary for UCCSD with an imaginary time evolution, additional tests are needed. But I think this original logic is also workaround, isn't it?

Reply:
I am not sure how the above change relates to 1311. The code above is in converting of a legacy WeightedPauliOperator ( which internally is a list (sum) of Paulis where each entry in the list is a complex coeff plus a Pauli. Although its complex I think currently it has a 0 imaginary part. The old code that evolved the operator is here where we needed to check and do some workaround with regard to Terra Parameters is here  https://github.com/Qiskit/qiskit-aqua/blob/091be3d38636352a39a0aecc310a73e8610b7bfb/qiskit/aqua/operators/legacy/common.py#L326-L331

As far as I can see, in the code you changed, it to still allows complex to pass through it unchanged if it actually has any non-zero imaginary part
```
>>> w = 1.2+0j
>>> np.real(w) if np.isreal(w) else w
1.2
>>> w = 1.2+8j
>>> np.real(w) if np.isreal(w) else w
(1.2+8j)
```
So your example of 1+1j - that the prior code produced 2, would let through 1+1j now. If that ends up in a Terra Parameter it will fail.

We cannot do imaginary time evolution. I cannot recall whether there is a case where the coeff had a non-zero imaginary part. The legacy operator though was made generic to hold a complex coeff for each Paul.

Reply:
Very sorry, not an issue 1311, but 1381.

Reply:
I think the complex coefficients in Legacy WeightedPauliOperator are rare and the purpose is limited. OpFlow allows complex coefficients, so each method in OpFlow and related modules must check the validity if the coeff is allowed to be a complex number or not.

Reply:
My type conversion is better from memory point of view.

Reply:
Ok, I guess this comment in the method doesn't help
```
            # Adding the imaginary is necessary to handle the imaginary coefficients in UCCSD.
            # TODO fix these or add support for them in Terra.
```
Here its more like you want to keep the complex coefficients now in this conversion but since often they are all real (even though the coefficients are of complex type, and areas of opflow cannot handle complex type at all, that making it real may allow more operators to be used. Things may break elsewhere due to allowing complex types though now but that is  a different problem. Since UCCSD is not yet converted we won't see issues from it, but if it were converted to opflow then the problem stated there would happen since the code now allows complex types to pass through if it cannot be represented as a real.

Maybe it would help to have some explanation in the text as to what is going on. I hope I am finally understanding things - thanks for your patience!